Fix JA4 fingerprinting

drnpkr opened the pull request:
According to the JA4 specification/description (line 2140), the ALPN field should contain the first and last characters of the first ALPN extension value.

Currently, nDPI uses the first and second characters (alpn[0] and alpn[1]), which produces incorrect fingerprints that do not match other JA4 implementations like Wireshark.

For example, with ALPN 'http/1.1':
- Current (incorrect): 'ht' (first + second char)
- Fixed (correct):     'h1' (first + last char)

This change ensures nDPI's JA4 implementation conforms to the official specification and maintains interoperability with other JA4 tools.

Fixes: Incorrect JA4 ALPN fingerprint generation

Please sign (check) the below before submitting the Pull Request:

- [ X] I have signed the ntop Contributor License Agreement at https://github.com/ntop/legal/blob/main/individual-contributor-licence-agreement.md
- [ X] I have read the contributing guidelines at https://github.com/ntop/nDPI/blob/dev/CONTRIBUTING.md
- [ ] I have updated the documentation (in doc/) to reflect the changes made (if applicable)

Link to the related [issue](https://github.com/ntop/nDPI/issues): https://github.com/ntop/nDPI/issues/2914

Describe changes:
  - **File**: `src/lib/protocols/tls.c`
  - **Line**: 2213
  - **Change**: `ja->client.alpn[1]` → `ja->client.alpn[strlen(ja->client.alpn)-1]`

drnpkr committed:
Fix JA4 ALPN fingerprint to use first and last characters

According to the JA4 specification (line 2139), the ALPN field should
contain the first and last characters of the first ALPN extension value.

Currently, nDPI uses the first and second characters (alpn[0] and alpn[1]),
which produces incorrect fingerprints that don't match other JA4
implementations like Wireshark.

For example, with ALPN 'http/1.1':
- Current (incorrect): 'ht' (first + second char)
- Fixed (correct):     'h1' (first + last char)

This change ensures nDPI's JA4 implementation conforms to the official
specification and maintains interoperability with other JA4 tools.

Fixes: Incorrect JA4 ALPN fingerprint generation

drnpkr committed:
Fix JA4 ALPN implementation to correctly parse first ALPN protocol

The previous fix attempted to use strlen(ja->client.alpn)-1 but this was
insufficient because nDPI modifies the ALPN string by:
1. Adding null terminators that truncate the last character
2. Converting semicolons to dashes, affecting multi-protocol ALPNs

This complete fix:
- Adds alpn_original_last field to store the true last character
- Captures the last character of the FIRST ALPN protocol only (before ;/,)
- Preserves the original character before nDPI's string modifications

Now correctly implements JA4 spec: first + last characters of first ALPN protocol
Examples:
- ALPN 'h2;http/1.1' -> 'h2' (not 'h.' or 'h1')
- ALPN 'http/1.1' -> 'h1' (not 'ht' or 'h.')

Fixes: #2914

drnpkr commented:
The previous fix attempted to use strlen(ja->client.alpn)-1 but this was
insufficient because nDPI modifies the ALPN string by:
-  Adding null terminators that truncate the last character
-  Converting semicolons to dashes, affecting multi-protocol ALPNs

The latest fix:
- Adds alpn_original_last field to store the true last character
- Captures the last character of the FIRST ALPN protocol only (before ;/,)
- Preserves the original character before nDPI's string modifications

Now correctly implements JA4 spec: first + last characters of first ALPN protocol
Examples:
- ALPN 'h2;http/1.1' -> 'h2' (not 'h.' or 'h1')
- ALPN 'http/1.1' -> 'h1' (not 'ht' or 'h.')

IvanNardi commented:
@drnpkr, could you update the unit tests results, please?
Something like `NDPI_FORCE_UPDATING_UTESTS_RESULTS=1 ./tests/do.sh` and push the changes...

drnpkr committed:
Update unit test results for JA4 fingerprint format changes

Updated test expectations to reflect the JA4 fingerprint format changes
from the ALPN bug fix. The changes primarily affect the JA4 hash format
in TLS and QUIC protocols.

drnpkr commented:
@IvanNardi sure thing! I already executed and pushed.

drnpkr commented:
@IvanNardi I also noticed one additional potential bug: I noticed that nDPI currently marks flows as having SNI (d) even when there is no actual TLS extension type 0x0000 (SNI) in the packet.

The technical details here https://github.com/FoxIO-LLC/ja4/blob/main/technical_details/JA4.md say: If the ClientHello contains an SNI extension (type `0x0000`), the SNI flag is `d`. Otherwise (no SNI), it is `i`. For some flows, currently nDPI incorrectly claims SNI present.

The bug is in line 2208 of src/lib/protocols/tls.c: https://github.com/drnpkr/nDPI/blob/6646cf113a43af6308dd7ce0289877d9cdc13523/src/lib/protocols/tls.c#L2208

`ja_str[3] = ndpi_isset_risk(flow, NDPI_NUMERIC_IP_HOST) ? 'i' : 'd', ja_str_len = 4;`

In my understanding:
  - nDPI uses NDPI_NUMERIC_IP_HOST risk to determine if SNI contains an IP address
  - If NDPI_NUMERIC_IP_HOST is set → `i` (IP address in SNI)
  - If NDPI_NUMERIC_IP_HOST is NOT set → `d` (assumes domain SNI present)

However, when there is NO SNI extension at all, NDPI_NUMERIC_IP_HOST is never set (it is only set when SNI exists AND contains an IP). Therefore, flows without any SNI extension incorrectly get `d` instead of `i`.

This is a relatively easy fix. This PR now fixes also https://github.com/ntop/nDPI/issues/2916

drnpkr committed:
Fix JA4 SNI detection to properly handle missing SNI extensions

Previously, nDPI incorrectly set JA4 SNI flag to 'd' (domain present) for
flows without any SNI extension. This was because the logic only checked
for NDPI_NUMERIC_IP_HOST risk (set when SNI contains IP) but didn't
distinguish between missing SNI and domain SNI.

Now properly detects:
- No SNI extension → 'i' flag
- SNI with IP address → 'i' flag
- SNI with domain → 'd' flag

This matches the JA4 specification.

Updated unit test results to reflect the corrected JA4 fingerprints.

IvanNardi reviewed (approved):
LGTM, thanks for these fixes!
